### ISSUE TITLE
ARM: Regression issue happened again since Mar-3-2016 (commit: 55a784c786e698dc394568fcdf6d42c9d0c827e0)

### ISSUE THREAD

**Issue by leemgs** [dotnet/coreclr#3545]
When we try to run "hello world" Raspberry Pi2 board  and Linux/ARM emulator since March-03-2016, 
the regression has again generated by some patches. First of all, I have tested my latest patches with commit Mar-3-2016 (commit: 55a784c786e698dc394568fcdf6d42c9d0c827e0). My patches is normal.

Anyone who has similar problem like me? Do I have this issue alone? 
CCed from Mar-3-2016 (commit: 55a784c786e698dc394568fcdf6d42c9d0c827e0) to Mar-5-2016 (commit: 8e069ce739232b43085e530556bf6b02832d8b5b) 

CC: @jkotas , @benpye
CC: @vcsjones, @JohnChen0 ,  @kyulee1, @ramarag, @briansull , @LLITCHEV , @gkhanna79 , @adityamandaleeka, @AndyAyersMS, @rahku , @adityamandaleeka, @myungjoo, @pgavlin, @willTG

I have summarized the my evaluation result as following: 

### Build environment:
```
- Build PC: Intel i5 CPU (quad core) , MEM 8GiB
- Build OS: Ubuntu 14.04 X86 64bit
- Memory: DDR3 8GiB
- Memory Cache: Not yet (e.g. ccache, distcache)
- build command: 
  u14.04$> cd ./coreclr
  u14.04$> time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  ./build.sh arm debug  clean cross  skiptests

```

### Running environment: 
```
- Target: Raspberry Pi2 board, Linux/ARM Emulator (for evaluating .NET Core)
- ARM CPU: ARM V7 architecture, Coretex-A9 (32bit)
- OS : Ubuntu/ARM OS (Ver. 14.04 Trusty Tahr)
- Link/Loader : glibc
- Default Floating Point: hard-fp(floating point) base

```

### Evaluation result for CoreCLR on Linux/ARM target: 
- Mar-3-2016 (commit: 55a784c786e698dc394568fcdf6d42c9d0c827e0) : 
  Build/Running is successful. 
- Test case of hello world (console app)
```
  root:/dotnet/runtime.debug.20160303> ./corerun  -c ./  example/hello.exe
   Hello World

```

- Mar-5-2016 (commit: 8e069ce739232b43085e530556bf6b02832d8b5b) :
  Build is OK. However, **Running is failed.**
- Test case of hello world (console app)
```
  root:/dotnet/runtime.debug.20160303> ./corerun  -c ./  example/hello.exe

Assert failure(PID 740 [0x000002e4], Thread: 740 [0x02e4]): Consistency check failed: Managed object size does not match unmanaged object size
man: 0xc, unman: 0x10, Name: System.Runtime.InteropServices.CriticalHandle
FAILED: size == expectedsize
    File: /work/dotnet/arm-32/coreclr/src/vm/binder.cpp Line: 538
    Image: /dotnet/runtime-debug-20160307/corerun

**** MessageBox invoked, title 'corerun - Assert Failure (PID 740, Thread 740/0x02e4)' ****
  Consistency check failed: Managed object size does not match unmanaged object size
man: 0xc, unman: 0x10, Name: System.Runtime.InteropServices.CriticalHandle
FAILED: size == expectedsize

/work/dotnet/arm-32/coreclr/src/vm/binder.cpp, Line: 538

Abort - Kill program
Retry - Debug
Ignore - Keep running


Image:
/dotnet/runtime-debug-20160307/corerun

********

Aborted
 root:/dotnet/runtime.debug.20160303> 
 root:/dotnet/runtime.debug.20160303> 
```

**Comment by janvorli** [comment]
@leemgs This error means that the mscorlib.dll doesn't match the libcoreclr.so. The need to come from the same version of the source tree and they need to be both debug or both release.

**Comment by leemgs** [comment]
@janvorli, It's strange. In my case, I have executed below command to build CoreCLR for Linux/ARM.
Actually, all generated files are located in same folder. Anyway, I will test with up-to-date commit no. (today's commit) again. 
* To build mscorlib: ./build.sh arm clean verbose 
* To build coreCLR: time ROOTFS_DIR=/work/dotnet/rootfs-coreclr/arm  ./build.sh arm debug  clean cross skipmscorlib skiptests


@benpye, Can you run "Hello Linux/ARM" console app on a Linux/ARM board with latest commit no.?

**Comment by janvorli** [comment]
@leemgs Looking at the CriticalHandle (which is referred to in the assert) definition in the libcoreclr and in the mscorlib and at the sizes reported by the assert, it really seems like a release version of mscorlib is used with debug version of libcoreclr.so. The CriticalHandle has 3 four-byte fields on ARM in release and 4 in debug, which would match the reported 0xC bytes on the mscorlib (man) and 0x10 bytes in the libcoreclr.so (unman) in this case.

**Comment by benpye** [comment]
@leemgs @janvorli I think you're right, looking at the build commands you
listed you omitted the debug parameter when building mscorlib.dll

**Comment by leemgs** [comment]
@janvorli , @benpye  I have found the reason of regression (e.g. running failure) that I reported. This reason is related to the #3529. I will send a PR after enough testing exception handling on build_mscorlib() in ./build.sh for Linux/ARM . 

Here is similar issues. 
a. "Failed to generate native image for mscorlib." happened when build mscorlib for Linux/ARM on Ubuntu 14.04 - https://github.com/dotnet/coreclr/issues/3529

b. Cannot handle exceptions properly (ARM/Linux/Debug build) -https://github.com/dotnet/coreclr/issues/3309
